Source/JavaScriptCore/wasm/WasmTypeDefinition.cpp

```diff
@@ -147,10 +209,100 @@ RefPtr<TypeDefinition> TypeDefinition::tryCreateStructType(StructFieldCount fiel
     void* memory = nullptr;
     if (!result.getValue(memory))
         return nullptr;
-    TypeDefinition* signature = new (NotNull, memory) TypeDefinition(fieldCount);
+    TypeDefinition* signature = new (NotNull, memory) TypeDefinition(TypeDefinitionKind::StructType, fieldCount);
+    return adoptRef(signature);
+}
+
+RefPtr<TypeDefinition> TypeDefinition::tryCreateRecursionGroup(RecursionGroupCount typeCount)
+{
+    // We use WTF_MAKE_FAST_ALLOCATED for this class.
+    auto result = tryFastMalloc(allocatedRecursionGroupSize(typeCount));
+    void* memory = nullptr;
+    if (!result.getValue(memory))
+        return nullptr;
+    TypeDefinition* signature = new (NotNull, memory) TypeDefinition(TypeDefinitionKind::RecursionGroup, typeCount);
     return adoptRef(signature);
 }
 
+RefPtr<TypeDefinition> TypeDefinition::tryCreateProjection()
+{
+    // We use WTF_MAKE_FAST_ALLOCATED for this class.
+    auto result = tryFastMalloc(allocatedProjectionSize());
+    void* memory = nullptr;
+    if (!result.getValue(memory))
+        return nullptr;
+    TypeDefinition* signature = new (NotNull, memory) TypeDefinition(TypeDefinitionKind::Projection);
+    return adoptRef(signature);
+}
+
+// Recursive types are stored "tied" in the sense that the spec refers to here:
+//
+//   https://github.com/WebAssembly/gc/blob/main/proposals/gc/MVP.md#equivalence
+//
+// That is, the recursive "back edges" are stored as a special type index. These
+// need to be substituted back out to a Projection eventually so that the type
+// can be further expanded if necessary. The substitute and unroll functions
+// below are used to implement this substitution.
+Type TypeDefinition::substitute(Type type, TypeIndex projectee)
+{
+    if (type.kind == TypeKind::Rec) {
+        RefPtr<TypeDefinition> projection = TypeInformation::typeDefinitionForProjection(projectee, static_cast<ProjectionIndex>(type.index));
+        TypeKind kind = type.isNullable() ? TypeKind::RefNull : TypeKind::Ref;
+        return Type { kind, type.nullable, projection->index() };
+    }
+
+    return type;
+}
+
+//
```

Comment:
Can we get a comment here?

Reply:
Added a comment here, and also renamed `unroll` to `replacePlaceholders` because this function didn't actually align very closely with what `unroll` is doing in the spec. The subtyping patch may end up adding an `unroll` again but for now it's not needed.